File: src/main/java/tn/esprit/nexaback/service/interfaces/FeedbackService.java
package tn.esprit.nexaback.service.interfaces;

import java.util.List;
import java.util.Map;

import tn.esprit.nexaback.persistance.entities.Feedback;


public interface FeedbackService {
	List<Feedback> retrieveAllFeedbacksByCourse(Integer id);
	Feedback createFeedback(String feedbackText, Integer userId, Integer courseId, Integer emotion);
	void removeFeedback(Integer id);
	Feedback updateFeedback(Integer id,String feedbackText, Integer emotion);
	Feedback toggleLike(Integer feedbackId, Integer userId);
	double getAverageEmotionForCourse(Integer courseId);
	Map<Integer, Map<String, List<Feedback>>> getFeedbacksGroupedByCourse();

}

File: src/app/pages/service/feedback.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

@Injectable({
  providedIn: 'root'
})
export class FeedbackService {
  private apiUrl = 'http://localhost:9090/api/feedbacks';

  constructor(private http: HttpClient) {}

  getFeedbacksByCourse(courseId: number): Observable<any[]> {
    return this.http.get<any[]>(`${this.apiUrl}/course/${courseId}`);
  }
  createFeedback(feedback: { feedbackText: string; userId: number; courseId: number; }): Observable<any> {
    return this.http.post<any>(`${this.apiUrl}/`, feedback);
  }
  toggleLike(feedbackId: number, userId: number): Observable<any> {
    return this.http.put<any>(`${this.apiUrl}/${feedbackId}/like/${userId}`, {});
  }
  getAverageEmotionForCourse(courseId: number): Observable<number> {
    return this.http.get<number>(`${this.apiUrl}/${courseId}/average-emotion`);
  }
  getGroupedFeedbacks(): Observable<{ [courseId: number]: { [courseTitle: string]: any[] } }> {
    return this.http.get<{ [courseId: number]: { [courseTitle: string]: any[] } }>(`${this.apiUrl}/groupedByCourse`);
  }
  deleteFeedback(feedbackId: number): Observable<any> {
    return this.http.delete<any>(`${this.apiUrl}/${feedbackId}`);
  }
}
